src/pow.rs

```diff
@@ -0,0 +1,1510 @@
+}
+
+impl Sub for Work {
+    type Output = Work;
+    fn sub(self, rhs: Self) -> Self { Work(self.0.wrapping_sub(rhs.0)) }
+}
+
+/// A 256 bit integer representing target.
+///
+/// The SHA-256 hash of a block's header must be lower than or equal to the current target for the
+/// block to be accepted by the network. The lower the target, the more difficult it is to generate
+/// a block. (See also [`Work`].)
+///
+/// ref: <https://en.bitcoin.it/wiki/Target>
+///
+/// # Examples
+///
+/// ```
+/// use bitcoin::Target;
+/// let _ = Target::from(0xdeadbeef_u32);
+/// ```
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+pub struct Target(U256);
+
+impl Target {
+    /// The maximum possible target.
+    ///
+    /// This value is used to calculate difficulty, which is defined as how difficult the current
+    /// target makes it to find a block relative to how difficult it would be at the highest
+    /// possible target (`max_target()`). Remember highest target == lowest difficulty.
+    ///
+    /// ref: <https://en.bitcoin.it/wiki/Target>
+    // In Bitcoind this is ~(u256)0 >> 32 stored as a floating-point type so it gets truncated, hence
+    // the low 208 bits are all zero.
+    pub fn max() -> Target { Target::from(0xFFFF_u64) << 208_u32 }
+
+    /// Computes the [`Target`] value from a compact representation.
+    ///
+    /// ref: <https://developer.bitcoin.org/reference/block_chain.html#target-nbits>
+    fn from_compact(bits: u32) -> Target {
+        // This is a floating-point "compact" encoding originally used by
+        // OpenSSL, which satoshi put into consensus code, so we're stuck
+        // with it. The exponent needs to have 3 subtracted from it, hence
+        // this goofy decoding code. 3 is due to 3 bytes in the mantissa.
+        let (mant, expt) = {
+            let unshifted_expt = bits >> 24;
+            if unshifted_expt <= 3 {
+                ((bits & 0xFFFFFF) >> (8 * (3 - unshifted_expt as usize)), 0)
+            } else {
+                (bits & 0xFFFFFF, 8 * ((bits >> 24) - 3))
+            }
+        };
+
+        // The mantissa is signed but may not be negative.
+        if mant > 0x7F_FFFF {
+            Target::default()
+        } else {
+            Target::from(mant as u64) << expt
+        }
+    }
+
+    /// Computes the compact value from a [`Target`] representation.
+    fn to_compact(self) -> u32 {
+        let mut size = (self.0.bits() + 7) / 8;
+        let mut compact = if size <= 3 {
+            (self.0.low_u64() << (8 * (3 - size))) as u32
+        } else {
+            let bn = self >> (8 * (size - 3));
+            bn.0.low_u32()
+        };
+
+        if (compact & 0x0080_0000) != 0 {
+            compact >>= 8;
+            size += 1;
+        }
+
+        compact | (size << 24) as u32
+    }
```

Comment:
Oh yes, I learned that during all the `to_u32` work and did not apply it here. Thanks.

Reply:
I made `to_compact` and `from_compact` both public, and use `CompactTarget` instead of `u32`. This means users now have available the following:
```rust
// let bits = <get these off the wire>;
let t = Target::from_compact(CompactTarget::from_consensus(bits));
let t: Target = CompactTarget::from_consensus(bits).into();
```

Reply:
In general is this idea of providing multiple ways of doing a thing something we want to support or should we stick to being opinionated and providing only a singe way (like we do for `to_consensus`). Or is it a case by case decision?

Reply:
It is generally useful to have at least `to_`/`into_` methods equivalent to the impl `Into<T>` for commonly used conversions to help inference.